From d62184c00829033d6aa798b4541a57922aced44f Mon Sep 17 00:00:00 2001
From: Ramya Gnanasekar <quic_rgnanase@quicinc.com>
Date: Tue, 24 Jan 2023 22:38:03 +0530
Subject: [PATCH] mac80211: RU puncturing support for STA

When STA associates with AP which advertises puncturing bitmap,
STA should also disable the channels punctured in AP and associate in
proper bandwidth.

This patch takes care of parsing EHT operational IE and updating
chandef with puncturing bitmap which in turn is passed to FW.

Signed-off-by: Ramya Gnanasekar <quic_rgnanase@quicinc.com>

Index: b/net/mac80211/ieee80211_i.h
===================================================================
--- a/net/mac80211/ieee80211_i.h	2023-03-06 14:11:55.984439895 +0530
+++ b/net/mac80211/ieee80211_i.h	2023-03-06 14:11:55.980439920 +0530
@@ -2470,6 +2470,10 @@ bool ieee80211_chandef_he_6ghz_oper(stru
 				    const struct ieee80211_he_operation *he_oper,
 				    const struct ieee80211_eht_operation *eht_oper,
 				    struct cfg80211_chan_def *chandef);
+bool ieee80211_chandef_eht_oper(struct ieee80211_sub_if_data *sdata,
+				const struct ieee80211_eht_operation *eht_oper,
+				bool support_160, bool support_320,
+				struct cfg80211_chan_def *chandef);
 bool ieee80211_chandef_s1g_oper(const struct ieee80211_s1g_oper_ie *oper,
 				struct cfg80211_chan_def *chandef);
 u32 ieee80211_chandef_downgrade(struct cfg80211_chan_def *c);
Index: b/net/mac80211/mlme.c
===================================================================
--- a/net/mac80211/mlme.c	2023-03-06 14:11:55.984439895 +0530
+++ b/net/mac80211/mlme.c	2023-03-06 14:12:19.116292421 +0530
@@ -157,6 +157,9 @@ ieee80211_determine_chantype(struct ieee
 	struct cfg80211_chan_def vht_chandef;
 	struct ieee80211_sta_ht_cap sta_ht_cap;
 	u32 ht_cfreq, ret;
+	const struct ieee80211_sta_eht_cap *eht_cap;
+	const struct ieee80211_sta_he_cap *he_cap;
+	enum nl80211_iftype iftype = ieee80211_vif_type_p2p(&sdata->vif);
 
 	memset(chandef, 0, sizeof(struct cfg80211_chan_def));
 	chandef->chan = channel;
@@ -247,7 +250,41 @@ ieee80211_determine_chantype(struct ieee
 	}
 
 	vht_chandef = *chandef;
-	if (!(ifmgd->flags & IEEE80211_STA_DISABLE_HE) && he_oper &&
+	if (!(ifmgd->flags & IEEE80211_STA_DISABLE_EHT) && eht_oper &&
+	      eht_oper->ehtop_params & EHTOP_PARAMS_EHTOP_INFORMATION_PRESENT) {
+		bool support_160, support_320;
+		u8 he_phy_cap, eht_phy_cap;
+
+		he_cap = ieee80211_get_he_iftype_cap(sband, iftype);
+		if (!he_cap) {
+			sdata_info(sdata, "Missing iftype sband data/HE cap, disabling HE/EHT");
+			ret = IEEE80211_STA_DISABLE_HE | IEEE80211_STA_DISABLE_EHT;
+			goto out;
+		}
+
+		eht_cap = ieee80211_get_eht_iftype_cap(sband, iftype);
+		if (!eht_cap) {
+			sdata_info(sdata, "Missing iftype sband data/EHT cap, disabling EHT");
+			ret = IEEE80211_STA_DISABLE_EHT;
+			goto out;
+		}
+
+		eht_phy_cap = eht_cap->eht_cap_elem.phy_cap_info[0];
+		he_phy_cap = he_cap->he_cap_elem.phy_cap_info[0];
+		support_320 =
+			eht_phy_cap & IEEE80211_EHT_PHY_CAP0_320MHZ_IN_6GHZ;
+		support_160 =
+			he_phy_cap &
+			IEEE80211_HE_PHY_CAP0_CHANNEL_WIDTH_SET_160MHZ_IN_5G;
+
+		if(!ieee80211_chandef_eht_oper(sdata, eht_oper, support_160,
+					       support_320, &vht_chandef)) {
+			sdata_info(sdata,
+				   "EHT AP information is invalid, disabling EHT");
+			ret = IEEE80211_STA_DISABLE_EHT;
+			goto out;
+                }
+	} else 	if (!(ifmgd->flags & IEEE80211_STA_DISABLE_HE) && he_oper &&
 	    (le32_to_cpu(he_oper->he_oper_params) &
 	     IEEE80211_HE_OPERATION_VHT_OPER_INFO)) {
 		struct ieee80211_vht_operation he_oper_vht_cap;
@@ -5676,6 +5713,7 @@ int ieee80211_mgd_assoc(struct ieee80211
 	const struct cfg80211_bss_ies *beacon_ies;
 	struct ieee80211_supported_band *sband;
 	struct ieee80211_bss_conf *bss_conf = &sdata->vif.bss_conf;
+	struct ieee80211_eht_operation *eht_oper;
 	const struct element *ssid_elem, *ht_elem, *vht_elem;
 	int i, err;
 	bool override = false;
@@ -5919,6 +5957,32 @@ int ieee80211_mgd_assoc(struct ieee80211
 	if (req->flags & ASSOC_REQ_DISABLE_EHT)
 		ifmgd->flags |= IEEE80211_STA_DISABLE_EHT;
 
+	/* TODO: To add QCN Vendor IE implemention for 240MHz
+	 * For now disable 320MHz support and downgrade to EHT160
+	 * in 5GHz
+	 */
+	if (is_5ghz) {
+		rcu_read_lock();
+		beacon_ies = rcu_dereference(req->bss->beacon_ies);
+		if (beacon_ies) {
+			const struct element *elem;
+			elem = cfg80211_find_ext_elem(WLAN_EID_EXT_EHT_OPERATION,
+						      beacon_ies->data, beacon_ies->len);
+			eht_oper = (const void *)(elem->data + 1);
+			if (elem && elem->datalen >= sizeof(*eht_oper) + 1 + 2 &&
+			    eht_oper-> ehtop_params &
+			    EHTOP_PARAMS_DISABLED_SUBCHANNEL_BITMAP_PRESENT) {
+				u8 bw = u8_get_bits(eht_oper->chan_width,
+						    IEEE80211_EHT_OPER_CHAN_WIDTH);
+				u16 bitmap = get_unaligned_le16(eht_oper->disable_subchannel_bitmap);
+				if (bw == IEEE80211_EHT_OPER_CHAN_WIDTH_320MHZ &&
+				   (bitmap & FIXED_PUNCTURE_PATTERN) == FIXED_PUNCTURE_PATTERN)
+					ifmgd->flags |= IEEE80211_STA_DISABLE_320MHZ;
+			}
+		}
+		rcu_read_unlock();
+	}
+
 	err = ieee80211_prep_connection(sdata, req->bss, true, override);
 	if (err)
 		goto err_clear;
Index: b/net/mac80211/util.c
===================================================================
--- a/net/mac80211/util.c	2023-03-06 14:11:55.984439895 +0530
+++ b/net/mac80211/util.c	2023-03-06 14:11:55.980439920 +0530
@@ -3525,7 +3525,7 @@ bool ieee80211_chandef_vht_oper(struct i
 	return true;
 }
 
-void ieee80211_chandef_eht_oper(struct ieee80211_sub_if_data *sdata,
+bool ieee80211_chandef_eht_oper(struct ieee80211_sub_if_data *sdata,
 				const struct ieee80211_eht_operation *eht_oper,
 				bool support_160, bool support_320,
 				struct cfg80211_chan_def *chandef)
@@ -3574,6 +3574,19 @@ void ieee80211_chandef_eht_oper(struct i
 		}
 		break;
 	}
+	if (eht_oper->chan_width >= IEEE80211_EHT_OPER_CHAN_WIDTH_80MHZ &&
+	    eht_oper->ehtop_params & EHTOP_PARAMS_DISABLED_SUBCHANNEL_BITMAP_PRESENT) {
+		chandef->ru_punct_bitmap = eht_oper->disable_subchannel_bitmap[0];
+
+		if (!cfg80211_ru_punct_bitmap_valid(chandef)) {
+			sdata_info(sdata, "Received invalid puncturing bitmap 0x%x",
+				   chandef->ru_punct_bitmap);
+			chandef->ru_punct_bitmap = 0;
+			return false;
+		}
+	}
+
+	return true;
 }
 
 
Index: b/include/net/cfg80211.h
===================================================================
--- a/include/net/cfg80211.h	2023-03-06 14:11:55.984439895 +0530
+++ b/include/net/cfg80211.h	2023-03-06 14:11:55.984439895 +0530
@@ -153,6 +153,7 @@ enum ieee80211_channel_flags {
 
 #define IEEE80211_DFS_MIN_CAC_TIME_MS		60000
 #define IEEE80211_DFS_MIN_NOP_TIME_MS		(30 * 60 * 1000)
+#define FIXED_PUNCTURE_PATTERN			0xF000
 
 /**
  * struct ieee80211_channel - channel definition
